Source/Core/Core/GeckoCode.cpp

```diff
@@ -156,35 +155,31 @@ static bool InstallCodeHandler()
 
 void RunCodeHandler()
 {
-  if (SConfig::GetInstance().bEnableCheats && active_codes.size() > 0)
-  {
-    if (!code_handler_installed || PowerPC::HostRead_U32(INSTALLER_BASE_ADDRESS) - 0xd01f1bad > 5)
-      code_handler_installed = InstallCodeHandler();
+  if (!SConfig::GetInstance().bEnableCheats)
+    return;
 
-    if (!code_handler_installed)
-    {
-      // A warning was already issued.
+  {
+    std::lock_guard<std::mutex> codes_lock(s_active_codes_lock);
+    if (s_active_codes.empty())
       return;
-    }
 
-    if (PC == LR)
+    if (!s_code_handler_installed || PowerPC::HostRead_U32(INSTALLER_BASE_ADDRESS) - 0xd01f1bad > 5)
     {
-      u32 oldLR = LR;
-      PowerPC::CoreMode oldMode = PowerPC::GetMode();
-
-      PC = INSTALLER_BASE_ADDRESS + 0xA8;
-      LR = 0;
-
-      // Execute the code handler in interpreter mode to track when it exits
-      PowerPC::SetMode(PowerPC::MODE_INTERPRETER);
+      s_code_handler_installed = InstallCodeHandlerLocked();
 
-      while (PC != 0)
-        PowerPC::SingleStep();
-
-      PowerPC::SetMode(oldMode);
-      PC = LR = oldLR;
+      // A warning was already issued for the install failing
+      if (!s_code_handler_installed)
+        return;
     }
   }
+
+  // If the last block that just executed ended with a BLR instruction then we can intercept it and
+  // redirect control into the Gecko Code Handler. The Code Handler will automatically BLR back to
+  // the original return address (which will still be in the link register) at the end.
+  if (PC == LR)
```

Comment:
But I'm pretty sure we are missing an `else` case here. The old behavior would only run the code handler *if and only if* this function is called after a block that has done a return. And this function is simply called called every 16.6ms, the chances of it actually being called after a blr are tiny.

So our once per frame code handler is only being cause a fraction of the time. I suspect this is the cause of the "Button codes" not working issue that @JMC47 has mentioned.